### PR TITLE
Forum

### PR DESCRIPTION
**Pull request by rmr167**
This pull request includes the basics for the forum feature of the website.  The main files added are ForumElement.java, ForumServlet.java, forum-script.js, and forum.html.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ForumServlet.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles forum data (questions, comments, and replies) */
+@WebServlet("/forum")
+public class ForumServlet extends HttpServlet {
+
+  /** Get request to the Datastore and responds with the forum elements with the parentId input */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long parentId = Long.parseLong(request.getParameter("parentId"));
```
Comment (nihabhaskar) [inline src/main/java/com/google/sps/servlets/ForumServlet.java:30]: parentId is the question itself right?

File: src/main/webapp/forum-script.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ *  Populates the replies for a forum element with the given id
+ *  in the placeholder given
+ *
+ *  @param {S.fn.init} placeholder the div that will hold the forum elements
+ *  @param {long} parentId the long that idntifies the parent of the forum
+ *  elements
+ */
+function expandForum(placeholder, parentId) {
```
Comment (nihabhaskar) [inline src/main/webapp/forum-script.js:18]: So this function would expand all the forum elements for a question when clicked?

Comment (nihabhaskar) [reply src/main/webapp/forum-script.js:18]: or does getReplies do that?

Comment (nihabhaskar) [reply src/main/webapp/forum-script.js:18]: I think something that could be helpful is adding function headers, for what each of these functions do.

**Comment by nihabhaskar** [comment]
Looks good! I just had a couple questions.

File: src/main/webapp/forum-script.js (hunk 2 above)
Comment (rmr167) [reply src/main/webapp/forum-script.js:20]: getReplies is called when a question's expand button is clicked.  However, it only determines the id and placeholder to send into expandForum and then calls expandForum.

File: src/main/java/com/google/sps/servlets/ForumServlet.java (hunk 1 above)
Comment (rmr167) [reply src/main/java/com/google/sps/servlets/ForumServlet.java:30]: Yeah that is a little confusing I will change that to be the id itself.  Thank you :)

Comment (rmr167) [reply src/main/java/com/google/sps/servlets/ForumServlet.java]: Updated the name to be id instead. Changed that also for doPost and for some functions in forum-script.js.  Much cleaner.

File: src/main/java/com/google/sps/data/ForumElement.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  private final long id;
+  private final long parentId;
+  private final String topic;
+  private final long timestamp;
+  private final long likes;
+  private final String text;
+  // private final long userId
+
+  public ForumElement(
+      long id, long parentId, String topic, long timestamp, long likes, String text) {
```
Comment (praneetmek) [inline src/main/java/com/google/sps/data/ForumElement.java:15]: I think likes and ids can just be ints. I think ids fit within int range, and i dont think we'll be getting anywhere near 2 billion likes (max bound of int in java)

File: src/main/java/com/google/sps/servlets/ForumServlet.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles forum data (questions, comments, and replies) */
+@WebServlet("/forum")
+public class ForumServlet extends HttpServlet {
+
+  /** Get request to the Datastore and responds with the forum elements with the parentId input */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long id = Long.parseLong(request.getParameter("id"));
```
Comment (praneetmek) [inline src/main/java/com/google/sps/servlets/ForumServlet.java:30]: Maybe you can keep id as a string? I did in mine but it doesnt make any difference

File: src/main/java/com/google/sps/data/ForumElement.java (hunk 3 above)
Comment (rmr167) [reply src/main/java/com/google/sps/data/ForumElement.java:15]: The id is originally outputted as a long not a String.  I think you may be thinking of the UserService id which is stored as a String.  So I will probably keep the id a long but I will try to change the likes to be int.

File: src/main/java/com/google/sps/servlets/ForumServlet.java (hunk 4 above)
Comment (rmr167) [reply src/main/java/com/google/sps/servlets/ForumServlet.java:30]: See above comment.

File: src/main/java/com/google/sps/data/ForumElement.java (hunk 3 above)
Comment (rmr167) [reply src/main/java/com/google/sps/data/ForumElement.java:15]: I store the likes as an int now.  Thanks for catching that.

**Pull request merged**